### ISSUE TITLE
PyTorch Autograd GIL Error in Deployed Flask App

### ISSUE THREAD

**Issue by arthurdjn** [pytorch/pytorch#47575]
## 🐛 Bug

Hello,
I am currently working on a small web application used to visualize CNN. 
My **Flask** application works perfectly in development mode, however when I deployed it I had an error using the `backward`  method:

```
RuntimeError: The autograd engine was called while holding the GIL. 
If you are using the C++ API, the autograd engine is an expensive operation that does not require the GIL to be held so you should release it with 'pybind11::gil_scoped_release no_gil;'. 
If you are not using the C++ API, please report a bug to the pytorch team.
```

This error always occurs when the `backward` method is called, even for small operations. I am using the python API (PyTorch).

## To Reproduce

I created my application with **Flask**, and deployed it using **Apache** and **wsgi**.

A similar application can be created in few lines, to illustrate my issue.

### Steps to reproduce the behavior:

The following tutorial to deploy the application on a ubuntu VPS is taken from [digital ocean](https://www.digitalocean.com/community/tutorials/how-to-deploy-a-flask-application-on-an-ubuntu-vps).

#### 1. Create a Flask application:

Create a directory `FlaskApp` inside `/var/www/`.
Then, create a second directory with the same name `FlaskApp`, and a file `flaskapp.wsgi`.
Inside the nested `FlaskApp`, create a `templates` and `static` folder, empty. Create a `__ init __.py` file too.

The tree should look like:

/var/www/FlaskApp/
├── FlaskApp
│   ├── __ init __ .py
│   ├── templates/
│   └── static/
└── flaskapp.wsgi

#### 2. FlaskApp / FlaskApp /__ init __.py

```python
from flask import Flask


app = Flask(__name__)


@app.route("/api/test")
def test():
    tensor1 = torch.tensor([1.], requires_grad=True)
    tensor2 = torch.tensor([2.], requires_grad=True)
    output= tensor1 * 10 + tensor2 * 2
    output.backward()

    return {
        "status": "done"
    }

if __name__ == "__main__":
    app.run(debug=True, host='127.0.0.1')
```

#### 3. FlaskApp / flaskapp.wsgi

```python
#!/usr/bin/python
import sys
import logging

logging.basicConfig(stream=sys.stderr)
sys.path.insert(0, "/var/www/FlaskApp/")

from FlaskApp import app as application
```

#### 4. Create an Apache conf

Create a `FlaskApp.conf` inside `/etc/apache2/sites-available/`:

```
<VirtualHost *:80>
		ServerName serverIP
		ServerAdmin admin@mywebsite.com
		WSGIScriptAlias / /var/www/FlaskApp/flaskapp.wsgi
		<Directory /var/www/FlaskApp/FlaskApp/>
			Order allow,deny
			Allow from all
		</Directory>
		Alias /static /var/www/FlaskApp/FlaskApp/static
		<Directory /var/www/FlaskApp/FlaskApp/static/>
			Order allow,deny
			Allow from all
		</Directory>
		ErrorLog ${APACHE_LOG_DIR}/error.log
		LogLevel warn
		CustomLog ${APACHE_LOG_DIR}/access.log combined
</VirtualHost>
```

#### 5. Test the server

The server should be configured, and can be executed with apache2.

Now, if you access to the route `http://serverIP/api/test`, the ``RunTimeError`` will be raised in the logs.

The full log is:

```
ERROR:FlaskApp.__init__:Exception on /api/test [GET]
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/dist-packages/flask/app.py", line 2447, in wsgi_app
    response = self.full_dispatch_request()
  File "/usr/local/lib/python3.8/dist-packages/flask/app.py", line 1952, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/usr/local/lib/python3.8/dist-packages/flask/app.py", line 1821, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/usr/local/lib/python3.8/dist-packages/flask/_compat.py", line 39, in reraise
    raise value
  File "/usr/local/lib/python3.8/dist-packages/flask/app.py", line 1950, in full_dispatch_request
    rv = self.dispatch_request()
  File "/usr/local/lib/python3.8/dist-packages/flask/app.py", line 1936, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/var/www/FlaskApp/FlaskApp/__init__.py", line 46, in test
    output.backward()
  File "/usr/local/lib/python3.8/dist-packages/torch/tensor.py", line 221, in backward
    torch.autograd.backward(self, gradient, retain_graph, create_graph)
  File "/usr/local/lib/python3.8/dist-packages/torch/autograd/__init__.py", line 130, in backward
    Variable._execution_engine.run_backward(
RuntimeError: The autograd engine was called while holding the GIL. If you are using the C++ API, the autograd engine is an expensive operation that does not require the GIL to be held so you should release it with 'pybind11::gil_scoped_release no_gil;'. If you are not using the C++ API, please report a bug to the pytorch team.
```

## Expected behavior

When the route `http://serverIP/api/test` is accessed, it should execute the `backward` method and return the response

```
{
    "status": "done"
}
```

## Environment

The details of my environment is (output from [collect_env script](https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py)):

Collecting environment information...
PyTorch version: 1.7.0+cpu
Is debug build: True
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.1 LTS (x86_64)
GCC version: (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
Clang version: Could not collect
CMake version: Could not collect

Python version: 3.8 (64-bit runtime)
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.4
[pip3] torch==1.7.0+cpu
[pip3] torchvision==0.8.1+cpu
[conda] Could not collect

## Additional context

My flask application is hosted on [Pulseheberg](https://pulseheberg.com/cloud/vps-education), on an ubuntu VPS. I have 2GB of RAM and 30GB Hard Drive.

**Comment by albanD** [comment]
FYI: duplicate of https://discuss.pytorch.org/t/using-pytorch-autograd-in-flask-app-gil-issues/102008
@arthurdjn there is no need to do double posts. We look both at the forum and issues. (use the forum for questions and issues for bug/feature requests).

**Comment by ezyang** [comment]
possibly relateed https://github.com/pytorch/pytorch/issues/42125

**Comment by kshitij12345** [comment]
Like @ezyang has linked, seems like it could be related to the way WSGI works with GIL.

I can successfully call the `/api/test` when using Flask Test Server.

@arthurdjn Can you please verify if this works for you as well.
```python
from flask import Flask
from flask import jsonify

import torch

app = Flask(__name__)


@app.route("/api/test")
def test():
    tensor1 = torch.tensor([1.], requires_grad=True)
    tensor2 = torch.tensor([2.], requires_grad=True)
    output= tensor1 * 10 + tensor2 * 2
    output.backward()

    return jsonify({
        "status": "done"
    })

if __name__ == "__main__":
    app.run(debug=False, host='127.0.0.1')
```

![Screenshot from 2020-11-10 15-35-54](https://user-images.githubusercontent.com/19503980/98659763-7b798880-236a-11eb-8877-b4ca2c32eea4.png)

**Comment by arthurdjn** [comment]
Thanks for all your answers.
The test server as highlighted by @kshitij12345 works. However, when I deploy it through WGSI the error is raised.
So I am not really sure if the error comes from PyTorch, or the deployment. Anyway, I will make other tests suggested by @albanD, @ezyang  and try to find what may cause this error. I will update the discussion when I will have more clues.

**Comment by arthurdjn** [comment]
I found my mistake. 
The issue was related to **WSGI** threads (and not **Flask** nor **PyTorch**), that's why the error occured only when I deployed my application.

For anyone who is facing the same problem, you can correct the virtual host configuration file. You will need to make sure that the application is running in the Python interpreter, read [`mod_wsgi` documentation](https://modwsgi.readthedocs.io/en/develop/user-guides/application-issues.html#python-simplified-gil-state-api) for more details or this related [issue on StackOverflow](https://stackoverflow.com/questions/49471240/slow-page-loading-on-apache-when-using-flask). 

In short, my configuration file is now:
```diff
<VirtualHost *:80>
	ServerName serverIP
	ServerAdmin admin@mywebsite.com
	WSGIScriptAlias / /var/www/FlaskApp/flaskapp.wsgi
	<Directory /var/www/FlaskApp/FlaskApp/>
		Order allow,deny
		Allow from all
	</Directory>
	Alias /static /var/www/FlaskApp/FlaskApp/static
	<Directory /var/www/FlaskApp/FlaskApp/static/>
		Order allow,deny
		Allow from all
	</Directory>
+       WSGIApplicationGroup %{GLOBAL}
	ErrorLog ${APACHE_LOG_DIR}/error.log
	LogLevel warn
	CustomLog ${APACHE_LOG_DIR}/access.log combined
</VirtualHost>
```

I am closing this issue, thanks again for your advices !

**Comment by laiviet** [comment]
> I found my mistake.
> The issue was related to **WSGI** threads (and not **Flask** nor **PyTorch**), that's why the error occured only when I deployed my application.
> 
> For anyone who is facing the same problem, you can correct the virtual host configuration file. You will need to make sure that the application is running in the Python interpreter, read [`mod_wsgi` documentation](https://modwsgi.readthedocs.io/en/develop/user-guides/application-issues.html#python-simplified-gil-state-api) for more details or this related [issue on StackOverflow](https://stackoverflow.com/questions/49471240/slow-page-loading-on-apache-when-using-flask).
> 
> In short, my configuration file is now:
> 
> ```diff
> <VirtualHost *:80>
> 	ServerName serverIP
> 	ServerAdmin admin@mywebsite.com
> 	WSGIScriptAlias / /var/www/FlaskApp/flaskapp.wsgi
> 	<Directory /var/www/FlaskApp/FlaskApp/>
> 		Order allow,deny
> 		Allow from all
> 	</Directory>
> 	Alias /static /var/www/FlaskApp/FlaskApp/static
> 	<Directory /var/www/FlaskApp/FlaskApp/static/>
> 		Order allow,deny
> 		Allow from all
> 	</Directory>
> +       WSGIApplicationGroup %{GLOBAL}
> 	ErrorLog ${APACHE_LOG_DIR}/error.log
> 	LogLevel warn
> 	CustomLog ${APACHE_LOG_DIR}/access.log combined
> </VirtualHost>
> ```
> 

This one works on my apache + django.
Thanks !
